ddtrace/contrib/botocore/patch.py

```diff
@@ -108,52 +109,78 @@ def inject_trace_data_to_message_attributes(trace_data, entry, endpoint=None):
-            entry["MessageAttributes"]["_datadog"] = {"DataType": "String", "StringValue": json.dumps(trace_data)}
+            entry["MessageAttributes"]["_datadog"] = {"DataType": "String", "StringValue": json.dumps(datadog_data)}
         elif endpoint == "sns":
             # Use Binary since SNS subscription filter policies fail silently
             # with JSON strings https://github.com/DataDog/datadog-lambda-js/pull/269
             # AWS will encode our value if it sees "Binary"
-            entry["MessageAttributes"]["_datadog"] = {"DataType": "Binary", "BinaryValue": json.dumps(trace_data)}
+            entry["MessageAttributes"]["_datadog"] = {"DataType": "Binary", "BinaryValue": json.dumps(datadog_data)}
         else:
             log.warning("skipping trace injection, endpoint is not SNS or SQS")
     else:
         # In the event a record has 10 or more msg attributes we cannot add our _datadog msg attribute
         log.warning("skipping trace injection, max number (10) of MessageAttributes exceeded")
 
 
-def inject_trace_to_sqs_or_sns_batch_message(params, span, endpoint=None):
+def set_checkpoint_and_get_datadog_data(pin, queue_name):
+    """
+    :pin: patch info for the botocore client
+    :queue_name: the name of the queue
+
+    Set the data streams monitoring checkpoint and return the encoded pathway
+    """
+    pathway = pin.tracer.data_streams_processor.set_checkpoint(["direction:out", "topic:" + queue_name, "type:sqs"])
+    return pathway.encode_b64()
+
+
+def inject_datadog_data_to_sqs_or_sns_batch_message(params, span, endpoint=None, pin=None, data_streams_enabled=False):
     # type: (Any, Span, Optional[str]) -> None
     """
     :params: contains the params for the current botocore action
     :span: the span which provides the trace context to be propagated
     :endpoint: endpoint of message, "sqs" or "sns"
+    :pin: patch info for the botocore client
+    :data_streams_enabled: boolean for whether data streams monitoring is enabled
 
-    Inject trace headers into MessageAttributes for all SQS or SNS records inside a batch
+    Inject trace headers and DSM info into MessageAttributes for all SQS or SNS records inside a batch
     """
-    trace_data = {}
-    HTTPPropagator.inject(span.context, trace_data)
+    if data_streams_enabled:
+        queue_url = params["QueueUrl"]
+        queue_name = queue_url[queue_url.rfind("/") + 1 :]
+
+    datadog_data = {}
+    HTTPPropagator.inject(span.context, datadog_data)
 
     # An entry here is an SNS or SQS record, and depending on how it was published,
     # it could either show up under Entries (in case of PutRecords),
     # or PublishBatchRequestEntries (in case of PublishBatch).
     entries = params.get("Entries", params.get("PublishBatchRequestEntries", []))
     for entry in entries:
-        inject_trace_data_to_message_attributes(trace_data, entry, endpoint)
+        if data_streams_enabled:
```

Comment:
Reverse the order here (DSM first, APM second), due to the scarcity of "message attributes."  If we hit the MessageAttribute limit, then enabling DSM will break tracing, which seems backwards since DSM is "Beta" for now